### DISCUSSION TITLE
The lifetime of a ref return should not be same to the ref struct itself

### DISCUSSION THREAD

**Discussion by 2A5F** [dotnet/csharplang#8556] [General]
```cs
public ref struct Foo1<T>(ref T[] arr)
{
    public ref T[] arr = ref arr;       // even T[] arr = arr; why?
    public ref T Ref() => ref arr[0];
}

public static class Foo2<T>
{
    public static ref T Some(T[] arr)
    {
        var a = new Foo1<T>(ref arr);
        return ref a.Ref();
                // ^ error CS8352: Cannot use variable 'a' in this context because it may expose referenced variables outside of their declaration scope
    }
}
```

`UnscopedRef` cannot be marked on the return value

https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKEIwGYACAJwFMAzMgZwQoFcBjBMgMQHtucAeACoA+ABTU6ggNoBdMhAoUAlIQDehMprKky0uQopkAvPMUBuDVp0TdZAEq1RS48Mq1TFKQAYZFggF9CYnIMHAA2bSwuXiwhYTVLTR1QiJtBMgBlbjAqUT0PFQItMnUi4q0ANwV5YzIYKgB3aL448XcDJT9yqwB2NzoIADoHGicurUCAwiA

---

> Note: Can bypass by Unsafe.AsRef

**Comment by jaredpar** [comment]
Lets track this out

```csharp
var a = new Foo1<T>(ref arr);
```

This is passing `ref arr` to a `ref struct` constructor. This is a local and the *ref-safe-context* of a local is the current method. That means that `a` has a *safe-context* of the current method. It cannot be returned. This in turn means that any _ref-like_ values that come from `a` are similarly restricted to the current method. This is why the compiler correctly errors on `return ref a.Ref()`.

In this particular case my question would be what is the reason for having `ref T[]` as an argument? That is the root of the complications here.

**Reply by RikkiGibson** [reply]
I thought that:
- `Foo1<T>.Ref()` cannot return a reference to `this`, so
- `a.Ref()` cannot return a reference to `a`, so
- `a.Ref()`'s *ref-safe-context* is *caller-context*.
- therefore `return ref a.Ref()` is OK.

**Reply by 2A5F** [reply]
> the reason for having ref T[] as an argument

This is a simplified example code, actually is ref to a struct that contains an array.

---

Here is another more simplified example.

```cs
public static class Foo<T>
{
    public static ref T Ref(ref T[] arr) => ref arr[0];
    
    public static ref T Some(T[] arr) => ref Ref(ref arr);
}

// error CS8166: Cannot return a parameter by reference 'arr' because it is not a ref parameter
// error CS8347: Cannot use a result of 'Foo<T>.Ref(ref T[])' in this context because it may expose variables referenced by parameter 'arr' outside of their declaration scope
```

Allowing UnscopedRef on return values ​​should solve this.

https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKEIwGYACDHANgqzIDEB7RgHgBUA+Qgb0LP4rlKNAE4BTAGZk2ZAEqSAFOKlsA2gF0yEESICUZALwcyyrTtUAGdQG4+Au/1IVqJydLIBlRmDEK1m7T1DY1N5CSU3QN1bAgBfQiA===

**Reply by jaredpar** [reply]
> Foo1<T>.Ref() cannot return a reference to this ...

It isn't returning a `ref` to `this` but insead a `ref` to a `ref` field. This particular case is [discussed](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/low-level-struct-improvements.md#provide-ref-fields-and-scoped) in the spec. This derives from how `Span<T>` worked from the beginning. It could always return a `ref` to a `ref` field but not to fields itself. 

> a.Ref()'s ref-safe-context is caller-context.

In the [rules for method invocation](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-11.0/low-level-struct-improvements.md#rules-for-method-invocation) this falls into the ref calling rules. Specifically that one of the minimums considered is ..

> The safe-context contributed by all argument expressions

The *safe-context* of `a` is _current method_ and hence so is the `ref` invocation.

**Reply by RikkiGibson** [reply]
> It could always return a ref to a ref field but not to fields itself.

Thanks, this clears it right up.

**Reply by jaredpar** [reply]
```csharp
public static ref T Ref(ref T[] arr) => ref arr[0];
```

In this case though the `ref` on the parameter is unnecessary. The `T[]` will always be on the heap so `ref` to the elements can be trivially returned. The use of `ref` isn't going to help the implementation of the method and will end up causing you headaches at the call site because the language will include the lifetime of the location for the array as part of the return lifetime calculation. 

Consider this example compiles just fine: 

```csharp
public static class Foo<T>
{
    public static ref T Ref(T[] arr) => ref arr[0];
    
    public static ref T Some(T[] arr) => ref Ref(arr);
}
```
[sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKEIwGYACDHANgqzIDEB7RgHgBUA+Qgb0LP4rlKNAE4BTAGZk2ZAEqSAFGwDaAXTIQRIgJRkAvBzLipmkcoAMqgNx8Bt/qQrUjk6WQDKjMGKVqNW3QMXKXkJBVNtGwIAX0IgA=)

**Reply by 2A5F** [reply]
The real code is usually more complex, and it may not be possible to directly use T[].

```cs
public interface IBox<T>
{
    public ref T Ref();
}

public struct Box<T> : IBox<T>
{
    // array or class, any $heap lifetime
    public T[] arr;
    // ... large struct
    
    public ref T Ref() => ref arr[0];
}

public static class Foo<T>
{
    public static ref T Ref<B>(ref B box) where B : IBox<T>
    {
        // ... some code
        return ref box.Ref();
    }
    
    public static ref T Some(T[] arr)
    {
        Box<T> box = new() { arr = arr };
        // ... some code
        ref var r = ref Ref(ref box);
        // ... some code
        return ref r;
    }
}
```
[sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKEIwGYACWBAUwCcAzCAYyrIEkAhAewA8AeAFQB8hAN6EyEsqTI0qdMvzIAlOQAoAlAG5CAX0LFyAZwQ0ArowRkufIWRBtrA4QTEFJZAPQeyEGjQgAnmScNFJYcD4wQQAkABZUEAAOZChQdFQIUGBU4pLS/ADaALo+ftpukl5kAHS1Kb4A5izGZha5Eu1S5LLyiip0GmQAvIIycqU0BQAMReV6RATSGDgAbGFkAGKcnE6inUurY73Kcrzsgqo9VmTAPOpkAO7xstf2HDy7FRKu7u5VtdUyIZONkyIxOAATHJfX4YADsRxuPGq/Q05Xc83c+3IyzWV0UAGUQVRVIUSr4aOpOj9fhJHLZbtxhmQYFQHoMRBNmRSyDp0bTPN4AUDiWDIdCBYiAG6+GTMq6oq6MrSdP5CurA0HgqGqvIIq40fkSeY6IA==)

<details>
  <summary>If have lifetime mark like rust</summary>

```cs
public interface IBox<'o, T>
{
    public ref<'o> T Ref();
}

public struct Box<T> : IBox<'heap, T>
{
    // array or class, any 'heap lifetime
    public T[] arr;
    // ... large struct
    
    public ref<'heap> T Ref() => ref arr[0];
}

public static class Foo<T>
{
    public static ref<'o> T Ref<'a, 'o, B>(ref<'a> B box) where B : IBox<'o, T>
    {
        // ... some code
        return ref box.Ref();
    }
    
    public static ref<'heap> T Some(T[] arr)
    {
        Box<T> box = new() { arr = arr };
        // ... some code
        ref var r = ref Ref(ref box);
        // ... some code
        return ref r;
    }
}
```

</details>

**Reply by jaredpar** [reply]
```csharp
public static ref T Ref<B>(ref B box) where B : IBox<T>
```

The problem with the sample is the `ref` on the parameter here. Because this is unannotated and the method is a `ref` return the compiler has to assume that it can participate in the return. Basically compiler has to assume that `ref T` could be pointing into `box`. 

This code sample does not do this though. Based on context I believe the `ref` passing style is to either to preserve mutations or for performance. Given that the `ref` is not returned as `ref` then have the compiler enforce that with `scoped`

```csharp
public static ref T Ref<B>(scoped ref B box) where B : IBox<T>
```

Once the compiler knows that the `ref` parameter can't be returned by `ref` then the sample compiles just fine:

[sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAIGYAEsECmATgGYQDGBOAkgEID2AHgDwAqAfALABQA3jzkE5sOIgRI5WOAEriAFAEoA3DwC+PHiIDOCIgFdyCHAxYccIGibZc+AoQHp7OCESIQAnjnpFhAJjjOMJ4AJAAWBBAADjgoUCQECFBgBHaCIqwA2gC6zq4q3EI4jjgAdGUxLgDmVDr6hqk4DSJiElKyJIo4ALzsouK5RBkADFn56tyauBgAjABsfjgAYvT01jz8BUIiM/MtkjLizLTsclrk9JEEACZ9ErQ4wEwKOADu4WLG5pZMa5uCG4VCsUyiUcFp6MkcOcrik/oCMAB2W4PJgldqKfKFcaFJpTObIqQAZQhBDkmRyLiICgaAMBgisZkejG6OBgBBenV4AxZlJwqkxdKKThBYJJUPoMIahT2ADcXKIWXt0XsmcopQ5heVwZDobDBcIkXsiALBONVEA===)

> If have lifetime mark like rust

That is roughly what `scoped` is doing here. It gives the lifetime of the `ref` the same lifetime as locals in the current method.

**Reply by 2A5F** [reply]
shit complex code, this is not a deliberately complex example, it is still a simplification of the actual code

```cs
public readonly ref struct RefBox<T>(ref T Ref)
{
    public readonly ref T Ref = ref Ref;
}

public interface IBox<out R> where R : allows ref struct
{
    public R Ref();
}

public ref struct Box<T, A> : IBox<RefBox<T>>
{
    public ref A data;
    // array or class, any $heap lifetime
    public T[] arr;
    // ... large struct
    
    public RefBox<T> Ref() => new(ref arr[0]);
}

public static class Foo<T>
{
    // if scoped
    // error CS8352: Cannot use variable 'scoped B box' in this context because it may expose referenced variables outside of their declaration scope
    public static R Ref<B, R>(B box) 
        where B : IBox<R>, allows ref struct
        where R : allows ref struct
    {
        // ... some code
        return box.Ref();
    }
    
    public static ref T Some<A>(T[] arr, A data)
    {
        Box<T, A> box = new() { arr = arr, data = ref data };
        // ... some code
        var r = Ref<Box<T, A>, RefBox<T>>(box);
        // ... some code
        return ref r.Ref;
        // error CS8352: Cannot use variable 'r' in this context because it may expose referenced variables outside of their declaration scope
    }
}
```

[godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIFCJsQAOpJfWQE8Ayo3QBhVLQCuLBiABspZwAZPAZMADkvACNMYhAAJn8LVAVCBwY3T28/UiSU%2BwFg0IiWaNiE60xbfIYhAnMCDK8ffxtMOzTa%2BsLwqJj4lrriBvcm7IVBgm7i0v6ASmtUD2Jkdg4LD0jaPGQAamJMJnQBWgBPPcwqHfHiDzsdgCULgCFUVQBSAGYXABVP7Ah9pdvg8LrM3hoAIJvADsT3BEJ2iJ260223Oh2OZ0BO2Bj0unwAIudLnjPnDITCCfD4SitrsQgQYlQmCsdgBJF7vL6LAgPP47ADuCBimAeOxAOzEtFQAoUxKuBBudmpsPhSORGzpYrxEDBH3JUOhVIpkNpaOx11uvM5nx%2BpB2UI%2B2HF7JtXzxbp%2Bf29FNVkPVZt22IR6CYdTJaqRAHoo5LiMQmGcSDtkAYFAp7YYzgASYVMCw7LZUTD2NiRxGBnFvACscOrRPMxAj/ujsYAdB3C%2BZgKLLcqW4jyxrUbsPa9bb8nSCqLqdoT%2BaEBQCLnGm7WNDWCXqDZTqabNWjxmG0ammOmdgAxVCoCc%2Bw0GgMH3ZH%2Byj6e2p72%2B5/CBPHaRV5ZjnAd1UFYV9h2P8JQ5cd3T%2BTNaGlWV5T7Agh3VIURTFCUpRlOULUVK0hxhB9QMRGMdg7NsrlQNgUwwTB0KRfYCCWBh/1eNsdW3YijWIkDKxfc0V2BIRaMYr5HX%2BX5a03Vd7RDMMmDBECSKYxFPW%2BBT%2BQA1Q5w%2BIlF1nEjV30ht43tUM6jM%2BUrKYOc%2BP1dSdgoqiaLotAsGcgA3cw9hs0kvk07SnS/Z5YK9J0f10niQPVVzOwUcT6K8uLmJLNj5WILiLmbBEkV3X1jQhDh5loThq14HwOC0UhUE4Fw3jiD4aKWVkmo%2BHhSAITRSvmABrEBqw0fROEkKrerqzheAUEARp6mrStIOBYCQNAWAsOgYnISh1s2%2BhYiOAhQgIQiGH6kwzEsNB9lTMgaFoRliFmiBIkmyIQnME5OC6j7mGIE4AHlIm0NoFq69a2EEQGGFOSasEiDxgBcKVZu4XgsBYQxgHERbSHwW77G8zA0dqzBVDaDxGR%2B3gGUqSatkiBMAbcLBJsVPAWBp0hieIAClAJTAsaMLYjF6%2BYqAMYAFAANTwTABUBixGG5/hBBEMR2CkGRBEUFR1Dx3QuH0bGUFMcwrEZ2bIHmVALGqNHeFQXniDwLBrd1CoqjSJwGFcEYsmNoIQh6Eo%2BmN3JUgERpA5yZIo4YKZeliY3WnaAROiGGOfFTyowY6CYk7DlPrAmbO9CPIYi5mLh5iStqtbKiqJrx%2BqOEosJsGBNx9hcQJ7h2CAsZCICIFwQhkw62veAWrR5ggJAeXWAgdogTB8CIV30D0NXhFEcRtd3vW1Emo3SAFBMLBppuOEq0hqtqtvAap5edlQS42077uSEwPuB6HpgI9B5uA2ltYgc5mrT26uLeYeYsCxE9oNYao0ODjXvpNNuM09Czz6iguILdH7TWgYtWY8xeYpEcJIIAA%3D%3D%3D)

(sharplab currently does not support net9

---

[actual code](https://godbolt.org/z/9T4eT4qK1)
[gist](https://gist.github.com/2A5F/2d7efd7b205771b628e35630880c54cd)

**Reply by jaredpar** [reply]
Think you can reduce that problem to the following code:

```csharp
ref struct S1; 
ref struct S2(ref int i)
{
    S1 Get() => default;

    S1 Test()
    {
        int i = 0;
        var s2 = new S2(ref i);
        return s2.Get();
    }
}
```

Essentially when a `ref struct` is returned from a `ref struct` the compiler assumes it must capture `ref` fields. This is a limitation of what you can state with the available lifetime features in C# today. 

In order to allow this pattern we'd need to effectively represent that the `this` inside an instance method can't escape any state. Basically we'd need to have [ref scoped](https://github.com/dotnet/csharplang/blob/main/proposals/expand-ref.md#ref-scoped-parameters) and make it work for the implicit `this` in a `struct` instance method. That would get the desired behavior here.

**Reply by 2A5F** [reply]
Will there be a proposal or something?

**Reply by jaredpar** [reply]
The `ref scoped` proposal I linked is what we would need here

https://github.com/dotnet/csharplang/blob/main/proposals/expand-ref.md

It doesn't explicitly cover this case but it's a fall out from the parameters work.
